A continuous time markov chain $X$ is defined by
$$\mathbb{P}(X_{t + dt} = j | X_t = i) = \delta_{ij} + q_{ij}dt$$
where $Q=(q_{ij})$ is an intensity matrix, that is
$$q_{ij} \geq 0,\ i\neq j$$
$$\sum_j q_{ij} = 0$$

In [2]:
from stochastic_process.markov_chain import MarkovChain

X = MarkovChain(1, 2, 5, 4, 6, 3)
X.Q

array([[-3.,  1.,  2.],
       [ 5., -9.,  4.],
       [ 6.,  3., -9.]])

Assuming $X$ is homogeneous in time, the matrix
$$P(\delta_t) = I + Q \delta_t$$
describe probabilities for $X$ to switch from one state to another within the $\delta_t$ time interval.

In [3]:
X.P(delta_t=0.02)

array([[0.94, 0.02, 0.04],
       [0.1 , 0.82, 0.08],
       [0.12, 0.06, 0.82]])

Note that $\delta_t$ must indeed be sufficiently small for $P(\delta_t)$ to be a probability matrix. For example $\delta_t = 0.2$ will be too large:

In [4]:
X.P(delta_t=0.2) # Return a matrix P with negative coefficients.

array([[ 0.4,  0.2,  0.4],
       [ 1. , -0.8,  0.8],
       [ 1.2,  0.6, -0.8]])

Now, given that $X$ is in a given state $i$, we generate a random number $r$ in the interval $[0, 1[$.
If $\delta_t = 0.02$, then 
$$P(\delta_t) = \left(\begin{aligned}
    0.94 && 0.02 && 0.04\\
    0.1 && 0.82 && 0.08\\
    0.12 && 0.06 && 0.82
\end{aligned}\right)$$
Assume that $X_t$ is in state 1, the probability for $X_{t+dt}$ to be in state
- 0 is 0.1
- 1 is 0.82
- 2 is 0.08

To simulate this, a number $r$ is randomly choose within the interval $[0,\ 1[$ and we consider the matrix $A$ defined by
$$ A = P(\delta_t)U =  \left(\begin{aligned}
    0.94 && 0.96 && 1\\
    0.1 && 0.92 && 1\\
    0.12 && 0.18 && 1
\end{aligned}\right)$$
where 
$$U = \left(\begin{aligned}
    1 && 1 && 1\\
    0 && 1 && 1\\
    0 && 0 && 1
\end{aligned}\right)$$ 

We then consider row 1 of $A$, for $X$^is in state 1, and replace each value by the boolean "$(r<A)$". For example, if $r = 0.3$ :
$$(0.1\ 0.92 \ 1) \longrightarrow (False,\ True,\ True)$$
then the next state of $X$ is the index of the first True value in "$(r<A)$".